src/clr-angular/forms/combobox/combobox.ts

```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+
     this.initializeSubscriptions();
   }
 
-  private initializeSubscriptions(): void {
-    this.subscription = this.optionSelectionService.renderSelectionChanged.subscribe((option: ClrOption<T>) => {
-      this.renderSelection(option);
-    });
+  get openState(): boolean {
+    return this.toggleService.open;
   }
 
-  private renderSelection(selectedOption: ClrOption<T>): void {
-    if (this.input && selectedOption) {
-      this.domAdapter.clearChildren(this.input.nativeElement);
-      const clone: HTMLElement = this.domAdapter.cloneNode(selectedOption.elRef.nativeElement);
-      this.renderer.setAttribute(clone, 'contenteditable', 'false');
-      this.renderer.appendChild(this.input.nativeElement, clone);
+  public smartPosition: ClrPopoverPosition = {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.START,
+    content: ClrAlignment.START,
+  };
+
+  loadingStateChange(state: ClrLoadingState): void {
+    this.optionSelectionService.loading = state === ClrLoadingState.LOADING;
+    this.positionService.realign();
+  }
+
+  private get disabled() {
+    return this.control && this.control.disabled;
+  }
+
+  unselect(item: T) {
+    if (!this.disabled) {
+      this.optionSelectionService.setSelected(item, false);
+      this.focusHandler.focus();
     }
   }
 
-  private registerPopoverIgnoredInput() {
-    if (this.input) {
-      // @TODO COMBOBOX: intentionally commented; resolve while merging the Combobox
-      //this.toggleService.registerIgnoredElement(this.input);
+  @Input('clrMulti')
+  set multiSelect(value: boolean) {
+    if (value) {
+      this.optionSelectionService.selectionModel = new MultiSelectComboboxModel<T>();
+    } else {
+      // in theory, setting this again should not cause errors even though we already set it in constructor,
+      // since the initial call to writeValue (caused by [ngModel] input) should happen after this
+      this.optionSelectionService.selectionModel = new SingleSelectComboboxModel<T>();
     }
   }
 
-  toggleOptionsMenu(event: MouseEvent): void {
-    this.toggleService.toggleWithEvent(event);
+  get multiSelect() {
+    return this.optionSelectionService.multi;
+  }
+
+  get ariaControls(): string {
+    return this.ariaService.ariaControls;
+  }
+
+  get ariaOwns(): string {
+    return this.ariaService.ariaOwns;
+  }
+
+  get ariaDescribedBySelection(): string {
+    return 'selection-' + this.id;
   }
 
-  @HostListener('click')
-  focusInput() {
-    if (this.input) {
-      this.domAdapter.focus(this.input.nativeElement);
+  get displayField(): string {
+    return this.optionSelectionService.displayField;
+  }
+
+  onFocus() {
+    this.domAdapter.addFocusClass(this.comboboxContainer.nativeElement);
+  }
+
+  onBlur() {
+    this.domAdapter.removeFocusClass(this.comboboxContainer.nativeElement);
+    this.onTouchedCallback();
+  }
+
+  @HostListener('keydown', ['$event'])
+  onKeyDown(event: KeyboardEvent) {
```

Comment:
Do we need the event argument here? I don't see it used anywhere.

Reply:
It's not used, but it is still supplied, I think. Unless the compiler complains, there's no overhead to leave it there, for maintenance purposes - better clarity what context is available in this handler. Or am I wrong? Do we gain anything by not specifying it?

Reply:
Thinking again about it, this is a common keydown handler. It should be a common knowledge how it's distributed. Removing.